File: solutions/Numbers/Prime_Factorization.js
#!/usr/bin/env node

/**
 * https://github.com/mkenney/Projects
 * Have the user enter a number and find all Prime Factors (if there are any)
 * and display them.
 */

var readline = require('readline');

/**
 * Return true if n is prime, else false
 */
function isPrime(n) {
	ret_val = true;
	if (2 != n) {
		if (2 > n || 0 == n % 2) {
			ret_val = false;

		} else {
			low_range = Math.sqrt(n) + 1;
			for (var a = 3; a < low_range; a += 2) {
				if (0 == n % a) {
					ret_val = false;
				}
			}
		}
	}
	return ret_val;
}

/**
 * Find the smallest factor of n
 */
function getSmallestFactor(n) {
	factor = false;
	if (n > 3) {
		factor_found = false;
		factor = 2;
		while (!factor_found) {
			if (0 == n % factor) {
				factor_found = true;
			} else {
				factor += 1;
			}
		}
	}
	return factor;
}

/**
 * Return a list of all the prime factors of n
 */
function getPrimeFactors(n) {
	factors = [];
	smallest = 0;
	cur_n = n;
	while (smallest != cur_n) {
		if (cur_n < 4 || isPrime(cur_n)){
			break;
		}

		smallest = getSmallestFactor(cur_n);
		factors.push(smallest);
		cur_n = cur_n / smallest;
	}

	if (cur_n > 1) {
		factors.push(cur_n);
	}

	return factors;
}

/**
 * Prompt for input and execute a callback on the result
 */
function prompt(question, callback) {
	var iface = readline.createInterface({
		input: process.stdin,
		output: process.stdout
	});
	iface.question(question, function(answer) {
		iface.close();
		callback(answer);
	});
};

/**
 * List prime factors until the user quits
 * @return {[type]} [description]
 */
function listPrimeFactors() {
	prompt("Enter an integer > 1 to to list it's prime factors ('q' to quit): ", function(n) {
		if ('q' != n) {
			n = Number(n);
			factors = getPrimeFactors(n);
			console.log("    Prime factors of "+n+": "+factors.join(', '));
			console.log('');
			listPrimeFactors();
		}
	});
}

listPrimeFactors();

File: solutions/Numbers/Prime_Factorization.py
#!/usr/bin/env python

"""
https://github.com/mkenney/Projects
Have the user enter a number and find all Prime Factors (if there are any)
and display them.
"""


def isPrime(n):
	"""
	Return True if n is prime, else False
	"""
	ret_val = True
	if 2 != n:
		if 2 > n or 0 == n % 2:
			ret_val = False

		else:
			low_range = int(n ** 0.5) + 1
			for a in range(3, low_range, 2):
				if 0 == n % a:
					ret_val = False;

	return ret_val


def getSmallestFactor(n):
	"""
	Find the smallest factor of n
	"""
	factor = False
	if n > 3:
		factor_found = False
		factor = 2
		while not factor_found:
			if 0 == n % factor:
				factor_found = True
			else:
				factor += 1

	return factor


def getPrimeFactors(n):
	"""
	Return a list of all the prime factors of n
	"""
	factors = []
	smallest = 0
	cur_n = n
	while smallest != cur_n:
		if cur_n < 4 or isPrime(cur_n):
			break

		smallest = getSmallestFactor(cur_n)
		factors.append(smallest)
		cur_n = cur_n / smallest

	if cur_n > 1:
		factors.append(cur_n)

	return factors

n = 0
while 'q' != n:
	n = raw_input("Enter an integer > 1 to to list it's prime factors ('q' to quit): ")
	if 'q' != n:
		n = int(n)
		factors = getPrimeFactors(n)
		print "    Prime factors of "+`n`+": "+', '.join(str(a) for a in factors)+"\n"
